### PR TITLE
[D&D] Adds drop validation

### PR DESCRIPTION
**Pull request by ashwin-pc**
### Description
This PR adds validation to check if the field being dragged can be dropped in a dropbox and if so which default aggregation should be used.
 
### Issues Resolved
-
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
    - [ ] `yarn test:jest`
    - [ ] `yarn test:jest_integration`
    - [ ] `yarn test:ftr`
- [ ] New functionality has been documented.
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx [hunk 1]
```diff
@@ -118,12 +122,18 @@ export const useDropbox = (props: UseDropboxProps): DropboxProps => {
 
   const onDropField = useCallback(
     (data: FieldDragDataType['value']) => {
-      if (!data) return;
+      if (!data || !validAggTypes.length) {
+        throw new Error('Cannot drop field here, missing drop properties');
+      }
```
Comment (joshuarrrr) [inline src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:127]: From a UX perspective, it seems like it would be better to determine the valid drop targets on the drag start event rather than waiting until the drop to declare it invalid. If we did that, we could highlight only valid drop targets, making it clear to the user where that field can go.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx [hunk 2]
```diff
@@ -118,12 +122,18 @@ export const useDropbox = (props: UseDropboxProps): DropboxProps => {
...
 
       const { name: fieldName } = data;
+      const schemaAggTypes = (schema.defaults as any).aggTypes;
+      const allowedAggTypes = schemaAggTypes
+        ? schemaAggTypes.filter((type) => validAggTypes.includes(type))
+        : [];
 
       aggConfigs?.createAggConfig({
-        type: (schema.defaults as any).aggType,
+        type: allowedAggTypes[0] || validAggTypes[0],
```
Comment (joshuarrrr) [inline src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:136]: I'm still catching up to some of the concepts in d&d, but why would we want to create an `aggConfig` even if there are no `allowedAggTypes` in the schema? Is it because schemas aren't required to define `allowedAggTypes`? If that's the case, why couldn't we just set L133 to `validAggTypes` and avoid the fallback here?

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx [hunk 3]
```diff
@@ -154,17 +164,36 @@ export const useDropbox = (props: UseDropboxProps): DropboxProps => {
...
+      if (!indexField) return [];
+
+      // Check if the aggTypes allowed by the schema allow this field
+      const aggTypes = aggService.types.getAll();
+
+      const allowedAggTypes = filterByName(aggTypes[schema.group], schema.aggFilter);
+
+      allowedAggTypes.forEach((aggType) => {
+        const allowedFieldTypes = aggType.paramByName('field')?.filterFieldTypes;
+        const isFieldDroppable = filterByType([indexField], allowedFieldTypes).length !== 0;
```
Comment (joshuarrrr) [inline src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:183]: nit - this name doesn't need to include `Droppable` - this test is more generic, something like `isFieldAllowed` or something like that. I also kind of think this would be more readable if written with `reduce()` instead of `forEach()`.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx [hunk 4]
```diff
@@ -154,17 +164,36 @@ export const useDropbox = (props: UseDropboxProps): DropboxProps => {
     onDropField
   );
 
-  const isValidDropField = useMemo(() => {
-    if (!dragData) return false;
+  useEffect(() => {
```
Comment (joshuarrrr) [inline src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:167]: This hook should probably be a separate file, partly because it looks like it will require substantial unit tests to verify all the edge cases.

**Review by joshuarrrr — APPROVED** [review]
Some questions and comments, but feel free to merge and keep the ball rolling.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx (hunk 1 above)
Comment (ashwin-pc) [reply src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:127]: Agreed. The user will never see this error. It will fail silently and only be logged in the console. The isValidDropTarget property returned by this hook is what actually indicates to the user that the field is droppable. And that is set when the drag begins. I just threw an error here for the console instead of failing silently if the user decided to ignore the UI indicators.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx (hunk 2 above)
Comment (ashwin-pc) [reply src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:136]: Great question. To understand the logic here, let me back up a bit since it took me a while to figure it out myself. On app start, the agg service registers a variety of aggregation types like avg, max, etc. But each schema has filters for the agg types they support and each aggtype has filters for the field types they support. So `validAggTypes` is an array of all the agg types that the field being dragged can be used with, and if there is atleast one of them, then we should be able to drop the field into that dropbox.

`allowedAggTypes` is a subset of `validAggTypes` where the aggTypes defined in `schema.defaults` are also present in `validAggTypes`. So the logic is that we use the default aggType thats also present in  `validAggTypes`, but if that isnt possible, we just use the first agg type from `validAggTypes` as a fallback.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx (hunk 3 above)
Comment (ashwin-pc) [reply src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx]: Not a fan of reduce, but i do agree with the readability. Ive refactored this to just use the filter and map functions.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx (hunk 4 above)
Comment (ashwin-pc) [reply src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx:165]: Good call. Ill leave it in here for this PR. I am yet to add another validation for min and max values. Once i do that, ill move it out with some tests.

File: src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx (hunk 1 above)
Comment (joshuarrrr) [reply src/plugins/wizard/public/application/contributions/containers/data_tab/items/use/use_dropbox.tsx]: Got it, that's helpful context.
